paho/client.go

```diff
@@ -127,154 +138,167 @@ func NewClient(conf ClientConfig) *Client {
+			debug.Println("received an error code in Connack:", ca.ReasonCode)
+			if ca.Properties != nil {
+				reason = ca.Properties.ReasonString
+			}
+			c.cerr = fmt.Errorf("failed to connect to server: %s", reason)
+			return
 		}
-		cleanup()
-		return ca, fmt.Errorf("failed to connect to server: %s", reason)
-	}
 
-	if ca.Properties != nil {
-		if ca.Properties.ServerKeepAlive != nil {
-			keepalive = *ca.Properties.ServerKeepAlive
-		}
-		if ca.Properties.AssignedClientID != "" {
-			c.ClientID = ca.Properties.AssignedClientID
-		}
-		if ca.Properties.ReceiveMaximum != nil {
-			c.serverProps.ReceiveMaximum = *ca.Properties.ReceiveMaximum
-		}
-		if ca.Properties.MaximumQoS != nil {
-			c.serverProps.MaximumQoS = *ca.Properties.MaximumQoS
-		}
-		if ca.Properties.MaximumPacketSize != nil {
-			c.serverProps.MaximumPacketSize = *ca.Properties.MaximumPacketSize
-		}
-		if ca.Properties.TopicAliasMaximum != nil {
-			c.serverProps.TopicAliasMaximum = *ca.Properties.TopicAliasMaximum
+		if ca.Properties != nil {
+			if ca.Properties.ServerKeepAlive != nil {
+				keepalive = *ca.Properties.ServerKeepAlive
+			}
+			//if ca.Properties.AssignedClientID != "" {
+			//	c.ClientID = ca.Properties.AssignedClientID
+			//}
+			if ca.Properties.ReceiveMaximum != nil {
+				c.serverProps.ReceiveMaximum = *ca.Properties.ReceiveMaximum
+			}
+			if ca.Properties.MaximumQoS != nil {
+				c.serverProps.MaximumQoS = *ca.Properties.MaximumQoS
+			}
+			if ca.Properties.MaximumPacketSize != nil {
+				c.serverProps.MaximumPacketSize = *ca.Properties.MaximumPacketSize
+			}
+			if ca.Properties.TopicAliasMaximum != nil {
+				c.serverProps.TopicAliasMaximum = *ca.Properties.TopicAliasMaximum
+			}
+			c.serverProps.RetainAvailable = ca.Properties.RetainAvailable
+			c.serverProps.WildcardSubAvailable = ca.Properties.WildcardSubAvailable
+			c.serverProps.SubIDAvailable = ca.Properties.SubIDAvailable
+			c.serverProps.SharedSubAvailable = ca.Properties.SharedSubAvailable
 		}
-		c.serverProps.RetainAvailable = ca.Properties.RetainAvailable
-		c.serverProps.WildcardSubAvailable = ca.Properties.WildcardSubAvailable
-		c.serverProps.SubIDAvailable = ca.Properties.SubIDAvailable
-		c.serverProps.SharedSubAvailable = ca.Properties.SharedSubAvailable
-	}
 
-	c.serverInflight = semaphore.NewWeighted(int64(c.serverProps.ReceiveMaximum))
-	c.clientInflight = semaphore.NewWeighted(int64(c.clientProps.ReceiveMaximum))
+		c.ca = ca
+		c.serverInflight = semaphore.NewWeighted(int64(c.serverProps.ReceiveMaximum))
+		c.clientInflight = semaphore.NewWeighted(int64(c.clientProps.ReceiveMaximum))
 
-	debug.Println("received CONNACK, starting pinger")
-	c.workers.Add(1)
-	go func() {
-		defer c.workers.Done()
-		c.pinger(time.Duration(keepalive) * time.Second)
-	}()
+		go c.pinger(time.Duration(keepalive) * time.Second)
+	})
+	return c.ca, c.cerr
+}
 
-	return ca, nil
+func (c *Client) waitConnected() {
+	var dummy bool
+	c.connectOnce.Do(func() {
+		dummy = true
+	})
+	if dummy {
```

Comment:
Could we potentially get a race by accessing `dummy` without locking here?

Reply:
No, it is protected by the `sync.Once`.

Reply:
OK, I was asking because I saw that `waitConnected` is called by a lot of methods and I haven't really checked whether those get called asynchronously or not.

Reply:
It's just a runtime check that connect was called before calling the method @fracasula